### ISSUE TITLE
`forge verify-bytecode`: full vs partial match regression

### ISSUE THREAD

**Issue by blmalone** [foundry-rs/foundry#8638]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (1197fbe 2024-08-09T00:25:08.996768000Z)

### What command(s) is the bug in?

forge verify-bytecode

### Operating System

macOS (Apple Silicon)

## Describe the issue
I suspect that a regression has been introduced in the `forge verify-bytecode` command. My understanding is that the implementation of this command is intended to adhere to the definitions of 'full' and 'partial' matches as outlined here: https://docs.sourcify.dev/docs/full-vs-partial-match/

In the Optimism monorepo, our contracts should _never_ have a full match. The reason `forge verify-bytecode` should never yield a 'full' match against Optimism contracts is because since tagged version [`op-contracts/v1.0.0`](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.0.0/packages/contracts-bedrock/foundry.toml#L16), we have always set the `foundry.toml` property [`bytecode_hash`](https://book.getfoundry.sh/reference/config/solidity-compiler?highlight=bytecode_hash#bytecode_hash) to `none`.

Which by definition means: 

> Determines the hash method for the metadata hash that is appended to the bytecode.

Therefore, the metadata hash is never appended to Optimism contracts.

### Current behavior (Full match)
```bash
# Update foundry to the latest with 'foundryup'
git clone https://github.com/ethereum-optimism/optimism.git
git checkout op-contracts/v1.3.0
cd optimism/packages/contracts-bedrock
rm -rf lib && forge clean && forge build --skip test contracts/test scripts contracts/dispute # skipping some contracts for speed here
forge verify-bytecode 0x2d778797049fe9259d947d1ed8e5442226dfb589 OptimismPortal --json
```

This returns:
```json
[   
    {"bytecode_type":"creation","match_type":"full"},
    {"bytecode_type":"runtime","match_type":"full"}
]
```

### Previous behavior (Partial match)
I built a nightly build from a few weeks ago at random: [here](https://github.com/foundry-rs/foundry/releases/tag/nightly-fe2acca4e379793539db80e032d76ffe0110298b) and was able to get a partial match. This means the regression has been introduced between now and 3 weeks ago.

```bash
# Update foundry to an older version of foundry without the regression
foundryup -b fe2acca4e379793539db80e032d76ffe0110298b 
git clone https://github.com/ethereum-optimism/optimism.git
git checkout op-contracts/v1.3.0
cd optimism/packages/contracts-bedrock
rm -rf lib && forge clean && forge build --skip test contracts/test scripts contracts/dispute # skipping some contracts for speed here
forge verify-bytecode 0x2d778797049fe9259d947d1ed8e5442226dfb589 OptimismPortal --json
```

This returns:
```json
[    
    {"bytecode_type":"creation","matched":true,"verification_type":"partial"}, 
    {"bytecode_type":"runtime","matched":true,"verification_type":"partial"}
]
```

## Testing
I've noticed several regressions while working with this tool. It would be beneficial to include tests alongside fixes and new features to catch regressions earlier in the development cycle. As this tool becomes more widely used, it is likely to become security-critical. False matches could have devastating downstream effects for teams.

Thank you.

**Comment by klkvr** [comment]
Right now `verify-bytecode` would always report "full" match in cases when on-chain bytecode exactly matches the one we have locally for the given contract. So, if both don't have metadata hash, this would be a "full" match. I think this makes sense as it is a bit confusing to receive "partial" matches when verifying contracts against local artifacts with exactly same configuration. However sourcify indeed reports "partial" matches when no metadata hash is appended, even when bytecodes are exactly same. I'm not opposed to changing this back, wdyt @yash-atreya @zerosnacks 

Regarding testing, we've recently added `verify-bytecode` tests and https://github.com/foundry-rs/foundry/pull/8510 will add more, so there should be less regressions after it

**Comment by yash-atreya** [comment]
> Right now `verify-bytecode` would always report "full" match in cases when on-chain bytecode exactly matches the one we have locally for the given contract. So, if both don't have metadata hash, this would be a "full" match. I think this makes sense as it is a bit confusing to receive "partial" matches when verifying contracts against local artifacts with exactly same configuration. However sourcify indeed reports "partial" matches when no metadata hash is appended, even when bytecodes are exactly same. I'm not opposed to changing this back, wdyt @yash-atreya @zerosnacks
> 
> Regarding testing, we've recently added `verify-bytecode` tests and #8510 will add more, so there should be less regressions after it

Yeah I'd like to revert to the previous definitions adhering to sourcify. Will do this as part of #8510

**Comment by yash-atreya** [comment]
> Yeah I'd like to revert to the previous definitions adhering to sourcify. Will do this as part of #8510

This has been addressed in #8510

**Comment by yash-atreya** [comment]
> @yash-atreya Additionally, when I pass a constructor arg to the `forge verify-bytecode` command, I don't see that constructor arg get appended to the initcode and embedded into the runtime code.
> 
> Perhaps with #8510 this is fixed?
> 
> **Steps to try it out for yourself:**
> 
> ```
> cd optimism
> git checkout a541c8a859d9258ad410598655f189de69adae19
> cd packages/contracts-bedrock
> nvm use 16 # need older version of node for yarn
> yarn install
> forge build --skip test contracts/test scripts contracts/dispute
> forge verify-bytecode 0x3d2c2f8f95CAba644eA25319c4c08594b8DC0359 OptimismMintableERC20Factory --json --constructor-args 0x3154cf16ccdb4c6d922629664174b904d80f2c35
> ```

Can confirm this is fixed in #8510.
